### ISSUE TITLE
Kinect Fusion fails on different OpenCL devices

### ISSUE THREAD

**Issue by cosiat** [opencv/opencv_contrib#2029]
##### System information (version)

- OpenCV => 4.0.1
- Operating System / Platform => Ubuntu 16.04 LTS
- Compiler => GCC 5.4.0

##### Detailed description

Different functions provided by the KinFu class fail depending on the used OpenCL device.

When using an Intel integrated GPU (the output of the `clinfo` command is in `clinfo.txt` in the attached 
[ZIP-archive](https://github.com/opencv/opencv_contrib/files/2937063/kinfu_opencl_issues.zip)), the `update` function of the KinFu object fails and returns false every time. This has been traced back to the OpenCL implementation of the `getAb` function in `src/fast_icp.cpp` (line 532). The result of the computation in the UMat `groupedSumGpu` seems to contain NaNs for an unknown reason, and the `upperTriangle` array fills with NaNs as a result (starting from line `src/fast_icp.cpp:609`).

When using an NVIDIA dedicated GPU, the `update` function works but the function `getPoints` of the KinFu object causes the program to crash with the following error message:
```
OpenCL error CL_OUT_OF_RESOURCES (-5) during call: clEnqueueNDRangeKernel('fillPtsNrm', dims=3, globalsize=512x512x512, localsize=8x8x16) sync=true
terminate called after throwing an instance of 'std::runtime_error'
  what():  Failed to run kernel
Aborted (core dumped)
```
With the OpenCL optimization disabled both functions work fine.

##### Steps to reproduce

Compile and install OpenCV 4.0.1 with `opencv_contrib` version tag 4.0.1, then unzip the [archive](https://github.com/opencv/opencv_contrib/files/2937063/kinfu_opencl_issues.zip) and compile the example replacing the `OpenCV_DIR` with the location of `OpenCVConfig.cmake` for OpenCV 4.0.1 if necessary:
```
unzip kinfu_opencl_issues.zip
cd kinfu_opencl_issues
cmake . -DOpenCV_DIR=/usr/local/lib/cmake/opencv4
make
./reproduce_kinfu_failures
```
Output using Intel GPU:
```
[ INFO:0] Initialize OpenCL runtime...
[ INFO:0] Successfully initialized OpenCL cache directory: /home/tkachuk/.cache/opencv/4.0/opencl_cache/
[ INFO:0] Preparing OpenCL cache configuration for context: Intel_R__Corporation--Intel_R__Gen9_HD_Graphics_NEO--19_03_12192
KinFu object ready
Update with frame 1: OK
Update with frame 2: FAIL
Update with frame 3: FAIL
Rendered image saved as rendered_image.png
Point cloud size: 11403
```
Output using NVIDIA GPU:
```
[ INFO:0] Initialize OpenCL runtime...
[ INFO:0] Successfully initialized OpenCL cache directory: /home/tkachuk/.cache/opencv/4.0/opencl_cache/
[ INFO:0] Preparing OpenCL cache configuration for context: NVIDIA_Corporation--GeForce_940MX--384_130
KinFu object ready
Update with frame 1: OK
Update with frame 2: OK
Update with frame 3: OK
Rendered image saved as rendered_image.png
OpenCL error CL_OUT_OF_RESOURCES (-5) during call: clEnqueueNDRangeKernel('fillPtsNrm', dims=3, globalsize=512x512x512, localsize=8x8x16) sync=true
terminate called after throwing an instance of 'std::runtime_error'
  what():  Failed to run kernel
Aborted (core dumped)
```
Output with disabled optimization (`setUseOptimized(false)` at `reproduce_kinfu_failures.cpp:15`):
```
KinFu object ready
Update with frame 1: OK
Update with frame 2: OK
Update with frame 3: OK
Rendered image saved as rendered_image.png
Point cloud size: 12642
```

**Comment by Jie2World** [comment]
I get same error when i run realsense2 SDK's rs-kinfu example. It uses the kinfu module of opencv4.0.1, which groupedSumGpu contains NaNs and then ICP function fail.   If add setUseOptimized(false), the reconstruction is slow, but not reset kinfu again.  The CPU optimization is default opened(the result of cv::useOptimized() is true). If disabled optimization, the program will use ICPImpl::getAb<Mat> instead of   ICPImpl::getAb<UMat> in src/fast_icp.cpp (the first function run on CPU, the second one run on GPU).

**Comment by Jie2World** [comment]
Do you solve this bug?

> ##### System information (version)
> * OpenCV => 4.0.1
> * Operating System / Platform => Ubuntu 16.04 LTS
> * Compiler => GCC 5.4.0
> 
> ##### Detailed description
> Different functions provided by the KinFu class fail depending on the used OpenCL device.
> 
> When using an Intel integrated GPU (the output of the `clinfo` command is in `clinfo.txt` in the attached
> [ZIP-archive](https://github.com/opencv/opencv_contrib/files/2937063/kinfu_opencl_issues.zip)), the `update` function of the KinFu object fails and returns false every time. This has been traced back to the OpenCL implementation of the `getAb` function in `src/fast_icp.cpp` (line 532). The result of the computation in the UMat `groupedSumGpu` seems to contain NaNs for an unknown reason, and the `upperTriangle` array fills with NaNs as a result (starting from line `src/fast_icp.cpp:609`).
> 
> When using an NVIDIA dedicated GPU, the `update` function works but the function `getPoints` of the KinFu object causes the program to crash with the following error message:
> 
> ```
> OpenCL error CL_OUT_OF_RESOURCES (-5) during call: clEnqueueNDRangeKernel('fillPtsNrm', dims=3, globalsize=512x512x512, localsize=8x8x16) sync=true
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  Failed to run kernel
> Aborted (core dumped)
> ```
> With the OpenCL optimization disabled both functions work fine.
> 
> ##### Steps to reproduce
> Compile and install OpenCV 4.0.1 with `opencv_contrib` version tag 4.0.1, then unzip the [archive](https://github.com/opencv/opencv_contrib/files/2937063/kinfu_opencl_issues.zip) and compile the example replacing the `OpenCV_DIR` with the location of `OpenCVConfig.cmake` for OpenCV 4.0.1 if necessary:
> 
> ```
> unzip kinfu_opencl_issues.zip
> cd kinfu_opencl_issues
> cmake . -DOpenCV_DIR=/usr/local/lib/cmake/opencv4
> make
> ./reproduce_kinfu_failures
> ```
> Output using Intel GPU:
> 
> ```
> [ INFO:0] Initialize OpenCL runtime...
> [ INFO:0] Successfully initialized OpenCL cache directory: /home/tkachuk/.cache/opencv/4.0/opencl_cache/
> [ INFO:0] Preparing OpenCL cache configuration for context: Intel_R__Corporation--Intel_R__Gen9_HD_Graphics_NEO--19_03_12192
> KinFu object ready
> Update with frame 1: OK
> Update with frame 2: FAIL
> Update with frame 3: FAIL
> Rendered image saved as rendered_image.png
> Point cloud size: 11403
> ```
> Output using NVIDIA GPU:
> 
> ```
> [ INFO:0] Initialize OpenCL runtime...
> [ INFO:0] Successfully initialized OpenCL cache directory: /home/tkachuk/.cache/opencv/4.0/opencl_cache/
> [ INFO:0] Preparing OpenCL cache configuration for context: NVIDIA_Corporation--GeForce_940MX--384_130
> KinFu object ready
> Update with frame 1: OK
> Update with frame 2: OK
> Update with frame 3: OK
> Rendered image saved as rendered_image.png
> OpenCL error CL_OUT_OF_RESOURCES (-5) during call: clEnqueueNDRangeKernel('fillPtsNrm', dims=3, globalsize=512x512x512, localsize=8x8x16) sync=true
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  Failed to run kernel
> Aborted (core dumped)
> ```
> Output with disabled optimization (`setUseOptimized(false)` at `reproduce_kinfu_failures.cpp:15`):
> 
> ```
> KinFu object ready
> Update with frame 1: OK
> Update with frame 2: OK
> Update with frame 3: OK
> Rendered image saved as rendered_image.png
> Point cloud size: 12642
> ```

Do you solve this bug?

**Comment by cosiat** [comment]
I have found a workaround to make the `update` function work on an Intel GPU, I just replaced the NaNs with zeroes at `fast_icp.cpp:619`:
```
    for(int y = 0; y < ngroups.height; y++)
    {
        const float* rowr = groupedSumCpu.ptr<float>(y);
        for(int x = 0; x < ngroups.width; x++)
        {
            const float* p = rowr + x*UTSIZE;
            for(int j = 0; j < UTSIZE; j++)
            {
                upperTriangle[j] += (cvIsNaN(p[j])) ? 0 : p[j]; // here
            }
        }
    }
```
Seems to be working correctly.

**Comment by Jie2World** [comment]
Thanks but sad. It is not working on a NVIDIA GPU.  The groupedSumCpu  not only contain nans, but also has inf. If disabled optimization, I meet another problem. While moving the camera, the result of getAb fuction, A is a zero matrix. The det of A also is zero and ICP  fail again. How can i repair the leaky bucket? Sad.

**Comment by savuor** [comment]
You may try this solution: #2106 
It works on all platforms I've tested with no performance decline.

If you still have problems you may try the following:
* update Ubuntu from 16.04 to 18.04 or switch to Windows 10
  - The only platform I've succeeded to reproduce this bug was `Ubuntu 16.04 + Intel iGPU`
  - Both `Ubuntu 18.04 + NVIDIA GPU` and `Windows 10 + Intel iGPU` work fine
* Disable GPU branch of ICP code. This imposes performance penalty on data transfer between host and GPU.
  - This can be done separately from other algorithms by commenting a piece of OpenCL-related code in `ICPImpl::estimateTransform()` method in file `fast_icp.cpp`
* If the bug also happens in `getPoints()` or `getNormals()` methods the whole GPU code should be disabled by calling `cv::useOptimized(false)`. This drops performance significantly.

**Comment by enjoybo** [comment]
> You may try this solution: #2106
> It works on all platforms I've tested with no performance decline.
> 
> If you still have problems you may try the following:
> 
> * update Ubuntu from 16.04 to 18.04 or switch to Windows 10
>   
>   * The only platform I've succeeded to reproduce this bug was `Ubuntu 16.04 + Intel iGPU`
>   * Both `Ubuntu 18.04 + NVIDIA GPU` and `Windows 10 + Intel iGPU` work fine
> * Disable GPU branch of ICP code. This imposes performance penalty on data transfer between host and GPU.
>   
>   * This can be done separately from other algorithms by commenting a piece of OpenCL-related code in `ICPImpl::estimateTransform()` method in file `fast_icp.cpp`
> * If the bug also happens in `getPoints()` or `getNormals()` methods the whole GPU code should be disabled by calling `cv::useOptimized(false)`. This drops performance significantly.

I tried this solution but the performance drops significantly.

**Comment by savuor** [comment]
@enjoybo What OpenCL platform do you use?

**Comment by carlos-martinez** [comment]
Hi,

I just wondering if this has been solved for NVIDIA cards. I am still having the same error mentioned before:

OpenCL error CL_OUT_OF_RESOURCES (-5) during call: clEnqueueNDRangeKernel('fillPtsNrm', dims=3, globalsize=512x512x512, localsize=8x8x16) sync=true

I have:

- OpenCV 4.2 (debug)
- geForce RTX 2060 SUPER on a desktop.  I update to the latest driver (44.219 - Studio Driver). The system had before Radeon Graphics card before

I used the code provided in thread. I tested the code  "reproduce_kinfu_failures_V2.cpp", which I just modified to print the extra information about the GPU device. I am attaching the source and output logs.

s there a solution besides disabling the optimization?

Thanks

PS ... I used tools (e.g. https://www.geeks3d.com/20200117/gpu-caps-viewer-1-44-0-released/) to verify that OpenCL was working in the GPU. 

[reproduce_kinfu_failures_v2.zip](https://github.com/opencv/opencv_contrib/files/4167806/reproduce_kinfu_failures_v2.zip)
